[lldb] Implement TypeSystemSwiftTypeRef::GetIndexOfChildMemberWithName

kastiglione opened the pull request:
This ports the implementation of `GetIndexOfChildMemberWithName`, from `SwiftASTContext` to `TypeSystemSwiftTypeRef`.

With this change, when compiling with swiftinterfaces, private properties in Swift types are now visible to lldb through the TypeRef metadata contained in the binary. Tests that expected private properties to not be accessible have been updated.

rdar://68171371

kastiglione committed:
thread exe_ctx through GetIndexOfChild(Member)?WithName

kastiglione committed:
call newly stubbed SwiftLanguageRuntime::GetIndexOfChildMemberWithName

kastiglione commented:
@adrian-prantl this doesn't need a review yet, but can you let me know what you think about the plumbing changes added in 
e5300ff and e80330f?

* e5300ff: Adds an `ExecutionContext *` parameter to all forms of `GetIndexOfChildMemberWithName`
* e80330f: Adds to `SwiftLanguageRuntime` a stub implementation of `GetIndexOfChildMemberWithName`, and calls that from `TypeSystemSwiftTypeRef` (including some manually handled validation)

In summary, all this is setup to implement and use `SwiftLanguageRuntimeImpl::GetIndexOfChildMemberWithName`.

If this is all fine, a follow up commit will add commits to implement the behavior.

adrian-prantl commented on lldb/include/lldb/Symbol/CompilerType.h:
```diff
@@ -324,7 +324,7 @@ class CompilerType {
   /// Lookup a child given a name. This function will match base class names and
   /// member member names in "clang_type" only, not descendants.
-  uint32_t GetIndexOfChildWithName(const char *name,
+  uint32_t GetIndexOfChildWithName(const char *name, ExecutionContext *exe_ctx,
```
We should eventually upstream this change, but I don't see an issue with doing so. @jimingham previously pointed out that requiring a _ValueObject_ in CompilerType is problematic, but there is prior art for requiring an optional ExecutionContext in calls like GetBitSize(). @jimingham do you agree with that characterization?

kastiglione replied:
for what it's worth, the functions I've seen need the process, so passing a process instead of `ExecutionContext` makes the dependency clearer, at the cost of being less flexible if in the future something else from the context is needed instead of or in addition to the process.

adrian-prantl replied:
Since this is an API for all TypeSystems, maybe keeping it more general is actually desirable? It also looks like most CompilerType methods (https://lldb.llvm.org/cpp_reference/classlldb__private_1_1CompilerType.html) take an _ExecutionContextScope_. Maybe we should also do that here too, for symmetry?

kastiglione replied:
The use cases and distinction between `ExecutionContext` and `ExecutionContextScope` is not clear to me at all. I read their docs and looked at their APIs. The primary reason I went with `ExecutionContext` here is because from `ValueObject` it's more direct/convenient.

Within `ValueObject`, both of these patterns exist:


```cpp
ExecutionContext exe_ctx(GetExecutionContextRef());
```

```cpp
ExecutionContext exe_ctx(GetExecutionContextRef());
someFunction(exe_ctx.GetBestExecutionContextScope());
```

What's unclear to me, is why bring `ExecutionContextScope` into the equation? From its documentation, `ExecutionContext` says:

> These objects are designed to be used for short term execution context object storage while a function might be trying to evaluate something that requires a thread or frame.

Which does describe the needs of `SwiftLanguageRuntime`.

jimingham replied:
> We should eventually upstream this change, but I don't see an issue with doing so. @jimingham previously pointed out that requiring a _ValueObject_ in CompilerType is problematic, but there is prior art for requiring an optional ExecutionContext in calls like GetBitSize(). @jimingham do you agree with that characterization?

That's right.  Since you really only want the process, or maybe a frame if you are trying to bind some generic, a ValueObject is way more than you need.  But that you might need a process to determine some aspects of the type system, it makes sense to have API's into the type system that provide an ExecutionContext in some form.

adrian-prantl replied:
> The use cases and distinction between ExecutionContext and ExecutionContextScope is not clear to me at all. I read their docs and looked at their APIs. 

Before I _guess_ an answer — @jimingham ^ ?

jimingham replied:
As I understand it, the ExecutionContextScope is there because it is really handy to be able to use one of the objects in the Execution Context hierarchy to stand in for a scope specifier regardless of where you are in the hierarchy.  I'm not sure for local passing of contexts there's a hard and fast rule on which to use, so far as I can tell it's more a matter of convenience.  If you know you will mostly always have a process/thread/frame object on hand when calling the API, then ExecutionContextScope is way more convenient as you can bypass the conversion to ExecutionContext.  

But you might also want to check with Greg in case I'm misremembering something.  These were his designs originally.

adrian-prantl replied:
Let's make the minimal API change to just add the context pointer here and then separately refurbish the API on llvm.org where we get rid of the char *, etc...

adrian-prantl commented on lldb/include/lldb/Target/SwiftLanguageRuntime.h:
```diff
@@ -256,6 +256,10 @@ class SwiftLanguageRuntime : public LanguageRuntime {
+  llvm::Optional<size_t> GetIndexOfChildMemberWithName(
+      CompilerType type, const char *name, ExecutionContext *exe_ctx,
+      bool omit_empty_base_classes, std::vector<uint32_t> &child_indexes);
+
```
This looks good — but it would be better to use an `llvm::StringRef` instead of a `const char*` for `name`.

adrian-prantl replied:
Similarly, a MutableArrayRef http://llvm.org/doxygen/classllvm_1_1MutableArrayRef.html is preferred over a std::vector&. Then call sites can also use SmallVectors if they wish to do so.

kastiglione replied:
thanks will do! TIL about `MutableArrayRef`

kastiglione replied:
done

kastiglione replied:
Seems `MutableArrayRef` won't work here, I don't see any way to modify an underlying vector, in particular to append to it.

kastiglione replied:
`MutableArrayRef` seems like a mutable _view_ over an immutable underlying `ArrayRef`.

adrian-prantl replied:
Whoops, you're right! We could go to a `SmallVectorImpl<uint32_t>` but the returns are marginal.

Side note: Having now read the documentation for `child_indexes` (https://lldb.llvm.org/cpp_reference/classlldb__private_1_1CompilerType.html#abbeabd29258512dc7869c3f01afff93d) this method's entire API is _weird_.

kastiglione replied:
haha very weird

adrian-prantl replied:
For posteriority: The point of the child_indexes array is to support returning more than one match. I have no idea in which language this would even be possible. To top that off, the return value is that of child_indexes[0] or INT_MAX, and the comment has a TODO mark in it. We might want to check if this is even wired up in _any_ language backend and otherwise potentially even remove it.

kastiglione replied:
sounds good to me.

In the case of multiple children with the same name, a class can have a property of the same name as a property in its base class if the base class property is private.

In my quick check of c++, it allows duplicate members names regardless of access control.

jimingham replied:
It really depends on what we decide to mean by "ChildMember".  If it means any member of the current object, then in C++ you can have two with the same name since they can come from different base classes.  If it only means "any member contributed by the most specific class of the object, and you have to get to the others by recursing through the base class by hand", then I can't think of a way you could have two members with the same name.

I think the first meaning is more useful, but I'm not 100% set on that.

adrian-prantl commented on lldb/source/Plugins/TypeSystem/Clang/TypeSystemClang.h:
```diff
@@ -832,7 +832,8 @@ class TypeSystemClang : public TypeSystem {
   // so we catch all names that match a given child name, not just the first.
   size_t
   GetIndexOfChildMemberWithName(lldb::opaque_compiler_type_t type,
-                                const char *name, bool omit_empty_base_classes,
```
When we upstream the change to add the exe_ctx, we can also modernize the signature to use StringRef and ArrayRef. And perhaps even return an Optional<>!

kastiglione replied:
💯

adrian-prantl commented on lldb/source/Plugins/TypeSystem/Swift/TypeSystemSwift.cpp:
```diff
@@ -101,3 +101,12 @@ bool TypeSystemSwift::ShouldTreatScalarValueAsAddress(
+  std::vector<uint32_t> child_indexes;
+  size_t num_child_indexes = GetIndexOfChildMemberWithName(
+      type, name, exe_ctx, omit_empty_base_classes, child_indexes);
+  return num_child_indexes == 1 ? child_indexes.front() : UINT32_MAX;
```
Oh yes. An Optional :-)

kastiglione committed:
use StringRef and MutableArrayRef in new API

kastiglione committed:
revert MutableArrayRef back to vector

kastiglione committed:
implement most of GetIndexOfChildMemberWithName

kastiglione commented:
@adrian-prantl d0a8d14 is the meat of this PR. Still to do:

- [x] Tune validation to handle cases where the runtime contains more info than the AST (0d5d9e2 and 93a4ce8)
- [x] Fix tests that assume private members are not valid (93a4ce8)
- [x] Debug failing `TestSwiftAtObjCIvars.py` test
- [ ] ~Implement `DumpTypeValue`~
- [x] Dereference weak types
- [x] Iterate over superclasses
- [ ] Update the swiftinterface tests to verify the contents of the newly visible members
- [x] Debug infinite(?) recursion in `TestSwiftProtocolTypes.py`

Note: there's now a lot of common structure (logic and flow control) in a few of these functions, which I am planning to refactor as a follow up.

kastiglione committed:
in GetChildCompilerTypeAtIndex, switch fallback() to impl()

Now that `TypeSystemSwiftTypeRef::GetNumChildren` is implemented, if the
runtime knows of more children than the AST contains, then the runtime should
be used (`impl`), not the AST (`fallback`).

kastiglione committed:
rework GetIndexOfChildMemberWithName validation

kastiglione committed:
first changes to tests now that these private members are visible

kastiglione commented:
`TestSwiftAtObjCIvars.py` is failing because it needs `DumpTypeValue` to be implemented using the runtime.

kastiglione committed:
take base class into account when calculating member index

kastiglione committed:
use <3 instead of <=2 - for consistency

kastiglione commented:
Follow up: `TestSwiftAtObjCIvars.py` failed because I had a buggy first version of b85f9cd.

kastiglione commented:
@swift-ci test

kastiglione committed:
dereference weak types

kastiglione committed:
traverse class hierarchy

kastiglione commented:
@adrian-prantl I think this is ready to review, the core changes are here 95dd368...ff5203d.

kastiglione committed:
fix superclass iteration

kastiglione committed:
return 0 not None for ThickFunction

kastiglione committed:
handle ExistentialMetatype in SwiftLanguageRuntimeImpl

adrian-prantl commented on lldb/source/Target/SwiftLanguageRuntimeDynamicTypeResolution.cpp:
```diff
@@ -1142,6 +1142,108 @@ GetTypeFromTypeRef(TypeSystemSwiftTypeRef &ts,
+  using namespace swift::reflection;
+  // Try the static type metadata.
+  const TypeRef *tr = nullptr;
+  auto *ti = GetTypeInfo(type, exe_ctx->GetFramePtr(), &tr);
```
this is outside of this patch, but maybe we should return a std::pair or a struct here.

kastiglione replied:
I was thinking that too. Initially, only one caller used the `TypeRef`, it was less disruptive to add as an optional out-param. But now more callers are using it, so I will make that follow up change.

adrian-prantl commented on lldb/source/Plugins/TypeSystem/Swift/TypeSystemSwiftTypeRef.cpp:
```diff
@@ -35,6 +35,9 @@
 #include "clang/APINotes/APINotesReader.h"
 
+#include <algorithm>
+#include <sstream>
```
We can avoid this...

adrian-prantl commented on lldb/source/Plugins/TypeSystem/Swift/TypeSystemSwiftTypeRef.cpp:
```diff
@@ -2365,11 +2368,70 @@ CompilerType TypeSystemSwiftTypeRef::GetChildCompilerTypeAtIndex(
+
+        auto fail = [&]() {
+          auto join = [](const auto &v) {
+            std::ostringstream buf;
```
... by calling llvm::join from StringExtras.h http://llvm.org/doxygen/namespacellvm.html#aa1fd1719afe5b07ebe55f1d105c98916

kastiglione replied:
I went to use `llvm::join`, but it's for containers of strings, and this case is a `vector<int>` basically. Since I'd need to convert to strings, it was shorter to do this, and since it will hopefully soon be deleted (that's the plan for the validation code right?), I didn't force the issue of using `llvm::join`.

adrian-prantl commented on lldb/source/Target/SwiftLanguageRuntimeDynamicTypeResolution.cpp:
```diff
@@ -1142,6 +1142,108 @@ GetTypeFromTypeRef(TypeSystemSwiftTypeRef &ts,
+    case ReferenceKind::Strong: {
+      auto *reflection_ctx = GetReflectionContext();
+      auto &builder = reflection_ctx->getBuilder();
+      auto tc = TypeConverter(builder);
```
`TypeConverter tc(builder)` otherwise this might call a pointless copy constructor?

kastiglione replied:
done

kastiglione committed:
update TestLibraryIndirect.py

adrian-prantl commented on lldb/test/API/lang/swift/parseable_interfaces/shared/TestSwiftInterfaceNoDebugInfo.py:
```diff
@@ -129,7 +129,7 @@ def do_test(self):
         lldbutil.check_variable(self, child_y, False, value="0")
 
         child_x = var.GetChildMemberWithName("x") # MyPoint.x isn't public
```
add `... but LLDB can find it through the type metadata`

kastiglione replied:
done

adrian-prantl reviewed (approved):
Great! Some minor suggestions inside.

kastiglione committed:
comment: LLDB can find it through type metadata

kastiglione committed:
use `T v(...)` style initialization

kastiglione commented:
@swift-ci test

kastiglione committed:
update No.swiftmodule.swift

kastiglione commented:
@swift-ci test

kastiglione commented:
@swift-ci test

kastiglione committed:
reenable TestSwiftFoundationTypeNotification.py

kastiglione committed:
add superclass to child_indexes

kastiglione commented:
@swift-ci test
